cpp/tests/ast/transform_tests.cpp

```diff
@@ -0,0 +1,281 @@
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <tests/utilities/base_fixture.hpp>
+#include <tests/utilities/column_utilities.hpp>
+#include <tests/utilities/column_wrapper.hpp>
+#include <tests/utilities/table_utilities.hpp>
+#include <tests/utilities/type_lists.hpp>
+
+#include <limits>
+#include <type_traits>
+
+template <typename T>
+using column_wrapper = cudf::test::fixed_width_column_wrapper<T>;
+
+struct ASTTest : public cudf::test::BaseFixture {
+};
+
+TEST_F(ASTTest, BasicAddition)
+{
+  auto c_0   = column_wrapper<int32_t>{3, 20, 1, 50};
+  auto c_1   = column_wrapper<int32_t>{10, 7, 20, 0};
+  auto table = cudf::table_view{{c_0, c_1}};
+
+  auto col_ref_0  = cudf::ast::column_reference(0);
+  auto col_ref_1  = cudf::ast::column_reference(1);
+  auto expression = cudf::ast::expression(cudf::ast::ast_operator::ADD, col_ref_0, col_ref_1);
+
+  auto expected = column_wrapper<int32_t>{13, 27, 21, 50};
+  auto result   = cudf::ast::compute_column(table, expression);
+
+  cudf::test::expect_columns_equal(expected, result->view(), true);
+}
+
+TEST_F(ASTTest, LessComparator)
+{
+  auto c_0   = column_wrapper<int32_t>{3, 20, 1, 50};
+  auto c_1   = column_wrapper<int32_t>{10, 7, 20, 0};
+  auto table = cudf::table_view{{c_0, c_1}};
+
+  auto col_ref_0  = cudf::ast::column_reference(0);
+  auto col_ref_1  = cudf::ast::column_reference(1);
+  auto expression = cudf::ast::expression(cudf::ast::ast_operator::LESS, col_ref_0, col_ref_1);
+
+  auto expected = column_wrapper<bool>{true, false, true, false};
+  auto result   = cudf::ast::compute_column(table, expression);
+
+  cudf::test::expect_columns_equal(expected, result->view(), true);
+}
+
+TEST_F(ASTTest, MultiLevelTreeArithmetic)
+{
+  auto c_0   = column_wrapper<int32_t>{3, 20, 1, 50};
+  auto c_1   = column_wrapper<int32_t>{10, 7, 20, 0};
+  auto c_2   = column_wrapper<int32_t>{-3, 66, 2, -99};
+  auto table = cudf::table_view{{c_0, c_1, c_2}};
+
+  auto col_ref_0 = cudf::ast::column_reference(0);
+  auto col_ref_1 = cudf::ast::column_reference(1);
+  auto col_ref_2 = cudf::ast::column_reference(2);
+
+  auto expression_left_subtree =
+    cudf::ast::expression(cudf::ast::ast_operator::ADD, col_ref_0, col_ref_1);
+
+  auto expression_right_subtree =
+    cudf::ast::expression(cudf::ast::ast_operator::SUB, col_ref_2, col_ref_0);
+
+  auto expression_tree = cudf::ast::expression(
+    cudf::ast::ast_operator::ADD, expression_left_subtree, expression_right_subtree);
+
+  auto result   = cudf::ast::compute_column(table, expression_tree);
+  auto expected = column_wrapper<int32_t>{7, 73, 22, -99};
+
+  cudf::test::expect_columns_equal(expected, result->view(), true);
+}
+
+TEST_F(ASTTest, ImbalancedTreeArithmetic)
+{
+  auto c_0   = column_wrapper<double>{0.15, 0.37, 4.2, 21.3};
```

Comment:
All of your tests operate on very small columns. I think there should be some tests that use large-ish columns (say 10K rows) because you have a custom kernel and bugs tend to creep in when you go from a single-block kernel to a multi-block kernel.

Also, do you have any tests that stress the shared memory limits? E.g. large number of operands?

Reply:
Wow. Sounds like something we don't have to worry about...

Reply:
Jake and I discussed and in order to move this forward, I'm willing to accept without the larger tests and file an issue to add larger tests in a follow-up.